### PR TITLE
Add automatic issue labeler

### LINKED ISSUE #1432: [mungegithub]  Create Munger To Automatically Add Labels To Issues
Issues currently require manual labeling for routing.   Let's automate the process with some simple machine learning.

### PR DESCRIPTION
**Pull request by grodrigues3**
Resolves #1432

The simple_app.py file loads 2 pretrained machine learning models that have learned patterns on existing issues.  For new issues, the munger file will send a request to the webapp with the new issue's title and body, and the webapp will respond with the team and component labels that it thinks are appropriate.

### CODE REVIEW

**Comment by grodrigues3** [comment]
@apelisse The simple app no longer depends on any of the other code that was used for model selection and validating the model.  It logs to a file (which maybe should be on the PV and it is currently storing the model on the container as well).  Let's change the configuration to point to the persistent volume.  Let me know if you see anything else that should be change (I can change it to a get request if you really want)

File: mungegithub/issue-labeler/simple_app.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+myPenalty = 'l2'
+myHasher = FeatureHasher(input_type="string", n_features= numFeatures, non_negative=True)
+myStemmer = PorterStemmer()
+tokenizer = RegexpTokenizer(r'\w+')
+
+
+try:
+  stop_words = []
+  if not stopwords:
+    stop_fn = "./data/stopwords.txt"
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: where is this file coming from?

File: mungegithub/issue-labeler/simple_app.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    comp_mod = joblib.load(component_fn)
+    vec = myHasher.transform([tokens])
+    tlabel = team_mod.predict(vec)[0]
+    clabel = comp_mod.predict(vec)[0]
+    return ",".join([tlabel, clabel])
+
+
+def tokenize_stem_stop(inputString):
+    curTitleBody = tokenizer.tokenize(inputString.decode('utf-8').lower())
+    return map(myStemmer.stem, filter(lambda x: x not in stopwords, curTitleBody))
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: how many stopwords do you have? Maybe you should use a `set` instead

File: mungegithub/issue-labeler/simple_app.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+  with open(stop_fn, 'r') as f:
+    stopwords = [word.strip() for word in f]
+except:
+  #don't remove any stopwords
+  stopwords = []
+
+@app.errorhandler(500)
+def internal_error(exception):
+  app.logger.error(exception)
+  return "Internal Server Error-> Check the Logs"
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: I would have done: `return str(exception)`. We already know it's an internal error (because of 500). And it's much simpler than checking the logs.

File: mungegithub/mungers/issue-categorizer.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
+/*
```
Comment (apelisse) [inline mungegithub/mungers/issue-categorizer.go]: Please apply some sort of `gofmt` to this file.

File: mungegithub/mungers/issue-categorizer.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	if obj.IsPR() {
+		return
+	}
+
+
+
+	issue := obj.Issue
+
+	existingLabels := issue.Labels
+	for _, label := range existingLabels{
```
Comment (apelisse) [inline mungegithub/mungers/issue-categorizer.go]: There is a function named `GetLabelsWithPrefix` in `mungegithub/github`:

```
github.GetLabelsWithPrefix(issue.Labels, 'team/')
github.GetLabelsWithPrefix(issue.Labels, 'component/')
```

File: mungegithub/mungers/issue-categorizer.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		}
+	}
+
+
+
+	routingLabelsToApply, err := http.PostForm("http://localhost:5000",
+		url.Values{"title": {*issue.Title}, "body": {*issue.Body}})
+
+	if routingLabelsToApply.StatusCode != 200{
+		glog.Error("Not 200 status code: " + string(routingLabelsToApply.StatusCode))
```
Comment (apelisse) [inline mungegithub/mungers/issue-categorizer.go]: You could print the error received from the server (the body)

File: mungegithub/issue-labeler/simple_app.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+def get_labels():
+    title = request.form.get('title', "")
+    body = request.form.get('body', "")
+    tokens = tokenize_stem_stop(" ".join([title, body]))
+    team_mod = joblib.load(team_fn)
+    comp_mod = joblib.load(component_fn)
+    vec = myHasher.transform([tokens])
+    tlabel = team_mod.predict(vec)[0]
+    clabel = comp_mod.predict(vec)[0]
+    return ",".join([tlabel, clabel])
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: This is not consistent with the string split below (`;` vs `,`)

File: mungegithub/issue-labeler/simple_app.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    return map(myStemmer.stem, filter(lambda x: x not in stopwords, curTitleBody))
+
+
+@app.route("/update_models", methods = ["PUT"])
+def update_model():
+    """
+    data should contain three fields
+      titles: list of titles
+      bodies: list of bodies
+      labels: list of list of labels
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: Not pretty.

More pretty:
```
data should be of the following form:
[
   {
      "title": "Issue Title",
      "body": "Issue body",
      "labels": ["team/something", "component/something"]
   },
   ...
]
```

File: mungegithub/issue-labeler/simple_app.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    titles = data.get('titles')
+    bodies = data.get('bodies')
+    labels = data.get('labels')
+
+    tTokens = []
+    cTokens = []
+    team_labels = []
+    component_labels = []
+    for (title, body, label_list) in zip(titles, bodies, labels):
+      tLabel = filter(lambda x: x.startswith('team'), label_list)
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: I find it interesting that you use filter rather than list comprehension. Whatever =)

File: mungegithub/issue-labeler/simple_app.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+      if tLabel:
+        team_labels += tLabel
+        tTokens += [tokens]
+      if cLabel:
+        component_labels += cLabel
+        cTokens += [tokens] 
+    tVec = myHasher.transform(tTokens)
+    cVec = myHasher.transform(cTokens)
+
+    if os.path.isfile(team_fn):
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: You could simplify the logic by switching if statements here and below:
```
if team_labels:
    if os.path.isfile(team_fn):
        ...
    else:
        ...
```

File: mungegithub/issue-labeler/simple_app.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+        component_model.partial_fit(cVec, np.array(component_labels))
+    else:
+      #create a new component model
+      if component_labels:
+        component_model = SGDClassifier(loss=myLoss, penalty=myPenalty, alpha=myAlpha) 
+        component_model.fit(cVec, np.array(component_labels))
+    
+    joblib.dump(team_model, team_fn)
+    joblib.dump(component_model, component_fn)
+    return "Nothing went wrong"
```
Comment (apelisse) [inline mungegithub/issue-labeler/simple_app.py]: You can return an empty string :-)

File: mungegithub/issue-labeler/simple_app.py (hunk 1 above)
Comment (grodrigues3) [inline mungegithub/issue-labeler/simple_app.py]: modified the dockerfile to add it to the container.  there are only 100ish words but I don't want to copy them all into the source file.

File: mungegithub/mungers/issue-categorizer.go (hunk 6 above)
Comment (grodrigues3) [inline mungegithub/mungers/issue-categorizer.go]: the line below this also checks if `err != nil`.   Is that redundant.  Does a non 200 status code automatically indicate there was an error and can I removed the second check.

File: mungegithub/issue-labeler/simple_app.py (hunk 8 above)
Comment (grodrigues3) [inline mungegithub/issue-labeler/simple_app.py]: this is for the comment right.  if so, i changed it for you.

File: mungegithub/issue-labeler/simple_app.py (hunk 9 above)
Comment (grodrigues3) [inline mungegithub/issue-labeler/simple_app.py]: interesting or bad? is it bad form for some reason?

File: mungegithub/submit-queue/deployment.yaml [hunk 12]
```diff
@@ -68,3 +68,18 @@ spec:
       - name: cache-volume
         persistentVolumeClaim:
           claimName: submit-queue-cache
+      - name: machine-learning-container
```
Comment (apelisse) [inline mungegithub/submit-queue/deployment.yaml]: Please revert this change!

File: mungegithub/issue-labeler/README.md [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+In order to run the webapp, build the container using the following command.
+
+$sudo docker build -no-cache -t <choose-image-name> . 
+
+Then, run the container making sure to forward traffic to port 5000.
+
+$docker run -p <port>:5000 <image name>
+
+
+
```
Comment (apelisse) [inline mungegithub/issue-labeler/README.md]: It misses a couple of blank lines I think ;-)

**Comment by apelisse** [comment]
Cool, can you please revert `deployment.yaml` change and squash these commits?
